src/components/flame-graph/Canvas.js

```diff
@@ -0,0 +1,239 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// @flow
+import * as React from 'react';
+import seedrandom from 'seedrandom';
+import withChartViewport from '../shared/chart/Viewport';
+import ChartCanvas from '../shared/chart/Canvas';
+import TextMeasurement from '../../utils/text-measurement';
+
+import type { Thread } from '../../types/profile';
+import type { CssPixels, UnitIntervalOfProfileRange } from '../../types/units';
+import type {
+  FlameGraphTiming,
+  FlameGraphDepth,
+  IndexIntoFlameGraphTiming,
+} from '../../profile-logic/flame-graph';
+import type { Action, ProfileSelection } from '../../types/actions';
+import type { CallNodeInfo } from '../../types/profile-derived';
+
+type Props = {
+  thread: Thread,
+  maxStackDepth: number,
+  containerWidth: CssPixels,
+  containerHeight: CssPixels,
+  viewportLeft: UnitIntervalOfProfileRange,
+  viewportRight: UnitIntervalOfProfileRange,
+  viewportTop: CssPixels,
+  viewportBottom: CssPixels,
+  flameGraphTiming: FlameGraphTiming,
+  callNodeInfo: CallNodeInfo,
+  stackFrameHeight: CssPixels,
+  updateProfileSelection: ProfileSelection => Action,
+  isDragging: boolean,
+};
+
+type HoveredStackTiming = {
+  depth: FlameGraphDepth,
+  flameGraphTimingIndex: IndexIntoFlameGraphTiming,
+};
+
+require('./Canvas.css');
+
+const ROW_HEIGHT = 16;
+const TEXT_OFFSET_START = 3;
+const TEXT_OFFSET_TOP = 11;
+
+function assignColor(depth, funcName) {
+  function hexStr(n) {
+    return ('0' + n.toString(16)).slice(-2);
+  }
+  // alea seems to be the fastest PRNG according to seedrandom README
+  const rng = seedrandom.alea(funcName + depth);
+  const red = hexStr(205 + parseInt(50 * rng()));
+  const green = hexStr(0 + parseInt(230 * rng()));
+  const blue = hexStr(0 + parseInt(55 * rng()));
+  return `#${red}${green}${blue}`;
+}
+
+class FlameGraphCanvas extends React.PureComponent<Props> {
+  _textMeasurement: null | TextMeasurement;
+
+  constructor(props: Props) {
+    super(props);
+    (this: any)._getHoveredStackInfo = this._getHoveredStackInfo.bind(this);
```

Comment:
We're now moving to using object initializers rather than `.bind(this)`. Right now we're in a mixed state of having both practices, but new could should follow the initializer approach.
e.g.

```js
class FlameGraphCanvas {
    // Remove constructor()
    
  _getHoveredStackInfo = _getHoveredStackInfo({
    depth,
    flameGraphTimingIndex,
  }: HoveredStackTiming): React.Node {
    // Function body here
  };

}
```

Let me know if you want me to elaborate on this.

Reply:
This sounds like my typing changes. I went ahead and rebased for you, which should fix up the weird issues.

https://github.com/gregtatum/perf.html/tree/brisad-flame-graph-canvas